### ISSUE TITLE
Can't mount singularity image in Vagrant Ubuntu box if in /vagrant

### ISSUE THREAD

**Issue by andersgs** [apptainer/singularity#3388]
## Version of Singularity:

3.1

### Expected behavior

Within the Ubuntu Vagrant Box in the `$HOME`

```
cd $HOME
singularity pull shub://GodloveD/lolcow
singularity --debut shell lolcow_latest.sif
```

```
VERBOSE [U=0,P=1042]       print()                       Set messagelevel to: 5
DEBUG   [U=0,P=1042]       init()                        PIPE_EXEC_FD value: 7
VERBOSE [U=0,P=1042]       init()                        Container runtime
VERBOSE [U=0,P=1042]       is_suid()                     Check if we are running as setuid
DEBUG   [U=0,P=1042]       init()                        Overlay seems supported by kernel
DEBUG   [U=0,P=1042]       init()                        Drop privileges
DEBUG   [U=900,P=1042]     init()                        Read json configuration from pipe
DEBUG   [U=900,P=1042]     init()                        Set child signal mask
DEBUG   [U=900,P=1042]     init()                        Wait completion of stage1
VERBOSE [U=900,P=1047]     priv_escalate()               Get root privileges
DEBUG   [U=0,P=1047]       set_parent_death_signal()     Set parent death signal to 9
DEBUG   [U=0,P=1047]       prepare_stage()               Entering in stage 1
DEBUG   [U=900,P=1047]     set_parent_death_signal()     Set parent death signal to 9
VERBOSE [U=900,P=1047]     init()                        Spawn stage 1
VERBOSE [U=900,P=1047]     startup()                     Execute stage 1
DEBUG   [U=900,P=1047]     Stage()                       Entering stage 1
DEBUG   [U=900,P=1047]     Init()                        Entering image format intializer
DEBUG   [U=900,P=1047]     Init()                        Check for image format sif
DEBUG   [U=900,P=1047]     Init()                        sif format initializer returns: SIF magic not found
DEBUG   [U=900,P=1047]     Init()                        Check for image format sandbox
DEBUG   [U=900,P=1047]     Init()                        sandbox format initializer returns: not a directory image
DEBUG   [U=900,P=1047]     Init()                        Check for image format squashfs
DEBUG   [U=900,P=1042]     init()                        Create socketpair for master communication channel
DEBUG   [U=900,P=1042]     cleanup_fd()                  Check file descriptor /proc/self/fd/3 pointing to /home/vagrant/lolcow_latest.sif
DEBUG   [U=900,P=1042]     cleanup_fd()                  Check file descriptor /proc/self/fd/4 pointing to anon_inode:[eventpoll]
DEBUG   [U=900,P=1042]     cleanup_fd()                  Closing /proc/self/fd/4
VERBOSE [U=900,P=1042]     priv_escalate()               Get root privileges
VERBOSE [U=0,P=1042]       create_namespace()            Create mount namespace
DEBUG   [U=0,P=1042]       init()                        Create RPC socketpair for communication between stage 2 and RPC server
DEBUG   [U=0,P=1042]       set_terminal_control()        Pass terminal control to child
VERBOSE [U=0,P=1042]       init()                        Spawn master process
DEBUG   [U=0,P=1053]       set_parent_death_signal()     Set parent death signal to 9
VERBOSE [U=0,P=1053]       create_namespace()            Create mount namespace
DEBUG   [U=0,P=1053]       set_parent_death_signal()     Set parent death signal to 9
DEBUG   [U=0,P=1053]       prepare_stage()               Entering in stage 2
DEBUG   [U=900,P=1053]     set_parent_death_signal()     Set parent death signal to 9
VERBOSE [U=0,P=1056]       init()                        Spawn RPC server
VERBOSE [U=900,P=1042]     startup()                     Execute master process
VERBOSE [U=0,P=1056]       startup()                     Serve RPC requests
DEBUG   [U=900,P=1042]     setupSessionLayout()          Attempting to use overlayfs (enable overlay = try)
DEBUG   [U=900,P=1042]     setupOverlayLayout()          Creating overlay SESSIONDIR layout
DEBUG   [U=900,P=1042]     addRootfsMount()              Mount rootfs in read-only mode
DEBUG   [U=900,P=1042]     addRootfsMount()              Mounting block [squashfs] image: /home/vagrant/lolcow_latest.sif
DEBUG   [U=900,P=1042]     addKernelMount()              Checking configuration file for 'mount proc'
DEBUG   [U=900,P=1042]     addKernelMount()              Adding proc to mount list
VERBOSE [U=900,P=1042]     addKernelMount()              Default mount: /proc:/proc
DEBUG   [U=900,P=1042]     addKernelMount()              Checking configuration file for 'mount sys'
DEBUG   [U=900,P=1042]     addKernelMount()              Adding sysfs to mount list
VERBOSE [U=900,P=1042]     addKernelMount()              Default mount: /sys:/sys
DEBUG   [U=900,P=1042]     addDevMount()                 Checking configuration file for 'mount dev'
DEBUG   [U=900,P=1042]     addDevMount()                 Adding dev to mount list
VERBOSE [U=900,P=1042]     addDevMount()                 Default mount: /dev:/dev
DEBUG   [U=900,P=1042]     addHostMount()                Not mounting host file systems per configuration
VERBOSE [U=900,P=1042]     addBindsMount()               Found 'bind path' = /etc/localtime, /etc/localtime
VERBOSE [U=900,P=1042]     addBindsMount()               Found 'bind path' = /etc/hosts, /etc/hosts
DEBUG   [U=900,P=1042]     addHomeStagingDir()           Staging home directory (/home/vagrant) at /usr/local/var/singularity/mnt/session/home/vagrant
DEBUG   [U=900,P=1042]     addHomeMount()                Adding home directory mount [/usr/local/var/singularity/mnt/session/home/vagrant:/home/vagrant] to list using layer: overlay
DEBUG   [U=900,P=1042]     isLayerEnabled()              Using Layer system: overlay
DEBUG   [U=900,P=1042]     addTmpMount()                 Checking for 'mount tmp' in configuration file
VERBOSE [U=900,P=1042]     addTmpMount()                 Default mount: /tmp:/tmp
VERBOSE [U=900,P=1042]     addTmpMount()                 Default mount: /var/tmp:/var/tmp
DEBUG   [U=900,P=1042]     addScratchMount()             Not mounting scratch directory: Not requested
VERBOSE [U=900,P=1042]     addCwdMount()                 Default mount: /home/vagrant: to the container
DEBUG   [U=900,P=1042]     addLibsMount()                Checking for 'user bind control' in configuration file
DEBUG   [U=900,P=1042]     addResolvConfMount()          Adding /etc/resolv.conf to mount list
VERBOSE [U=900,P=1042]     addResolvConfMount()          Default mount: /etc/resolv.conf:/etc/resolv.conf
DEBUG   [U=900,P=1042]     addHostnameMount()            Skipping hostname mount, not virtualizing UTS namespace on user request
DEBUG   [U=900,P=1042]     create()                      Mount all
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting tmpfs to /usr/local/var/singularity/mnt/session
DEBUG   [U=900,P=1042]     mountImage()                  Mounting loop device /dev/loop0 to /usr/local/var/singularity/mnt/session/rootfs
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting overlay to /usr/local/var/singularity/mnt/session/final
DEBUG   [U=900,P=1042]     setPropagationMount()         Set RPC mount propagation flag to SLAVE
VERBOSE [U=900,P=1042]     Passwd()                      Checking for template passwd file: /usr/local/var/singularity/mnt/session/rootfs/etc/passwd
VERBOSE [U=900,P=1042]     Passwd()                      Creating passwd content
VERBOSE [U=900,P=1042]     Passwd()                      Creating template passwd file and appending user data: /usr/local/var/singularity/mnt/session/rootfs/etc/passwd
DEBUG   [U=900,P=1042]     addIdentityMount()            Adding /etc/passwd to mount list
VERBOSE [U=900,P=1042]     addIdentityMount()            Default mount: /etc/passwd:/etc/passwd
VERBOSE [U=900,P=1042]     Group()                       Checking for template group file: /usr/local/var/singularity/mnt/session/rootfs/etc/group
VERBOSE [U=900,P=1042]     Group()                       Creating group content
DEBUG   [U=900,P=1042]     addIdentityMount()            Adding /etc/group to mount list
VERBOSE [U=900,P=1042]     addIdentityMount()            Default mount: /etc/group:/etc/group
DEBUG   [U=900,P=1042]     mountGeneric()                Remounting /usr/local/var/singularity/mnt/session/final
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /dev to /usr/local/var/singularity/mnt/session/final/dev
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /etc/localtime to /usr/local/var/singularity/mnt/session/final/etc/localtime
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /etc/hosts to /usr/local/var/singularity/mnt/session/final/etc/hosts
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /usr/local/etc/singularity/actions to /usr/local/var/singularity/mnt/session/final/.singularity.d/actions
DEBUG   [U=900,P=1042]     mountGeneric()                Remounting /usr/local/var/singularity/mnt/session/final/.singularity.d/actions
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /proc to /usr/local/var/singularity/mnt/session/final/proc
DEBUG   [U=900,P=1042]     mountGeneric()                Remounting /usr/local/var/singularity/mnt/session/final/proc
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting sysfs to /usr/local/var/singularity/mnt/session/final/sys
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /home/vagrant to /usr/local/var/singularity/mnt/session/home/vagrant
DEBUG   [U=900,P=1042]     mountGeneric()                Remounting /usr/local/var/singularity/mnt/session/home/vagrant
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /usr/local/var/singularity/mnt/session/home/vagrant to /usr/local/var/singularity/mnt/session/final/home/vagrant
DEBUG   [U=900,P=1042]     mountGeneric()                Remounting /usr/local/var/singularity/mnt/session/final/home/vagrant
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /tmp to /usr/local/var/singularity/mnt/session/final/tmp
DEBUG   [U=900,P=1042]     mountGeneric()                Remounting /usr/local/var/singularity/mnt/session/final/tmp
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /var/tmp to /usr/local/var/singularity/mnt/session/final/var/tmp
DEBUG   [U=900,P=1042]     mountGeneric()                Remounting /usr/local/var/singularity/mnt/session/final/var/tmp
DEBUG   [U=900,P=1042]     mountGeneric()                Skipping mount /usr/local/var/singularity/mnt/session/final/home/vagrant, /usr/local/var/singularity/mnt/session/final/home/vagrant already mounted
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /usr/local/var/singularity/mnt/session/etc/resolv.conf to /usr/local/var/singularity/mnt/session/final/etc/resolv.conf
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /usr/local/var/singularity/mnt/session/etc/passwd to /usr/local/var/singularity/mnt/session/final/etc/passwd
DEBUG   [U=900,P=1042]     mountGeneric()                Mounting /usr/local/var/singularity/mnt/session/etc/group to /usr/local/var/singularity/mnt/session/final/etc/group
DEBUG   [U=900,P=1042]     create()                      Chroot into /usr/local/var/singularity/mnt/session/final
DEBUG   [U=0,P=1056]       Chroot()                      Change current directory to /usr/local/var/singularity/mnt/session/final
DEBUG   [U=0,P=1056]       Chroot()                      Hold reference to host / directory
DEBUG   [U=0,P=1056]       Chroot()                      Called pivot_root on /usr/local/var/singularity/mnt/session/final
DEBUG   [U=0,P=1056]       Chroot()                      Change current directory to host / directory
DEBUG   [U=0,P=1056]       Chroot()                      Apply slave mount propagation for host / directory
DEBUG   [U=0,P=1056]       Chroot()                      Called unmount(/, syscall.MNT_DETACH)
DEBUG   [U=0,P=1056]       Chroot()                      Changing directory to / to avoid getpwd issues
DEBUG   [U=900,P=1042]     create()                      Chdir into / to avoid errors
VERBOSE [U=900,P=1053]     startup()                     Execute stage 2
DEBUG   [U=900,P=1053]     Stage()                       Entering stage 2
DEBUG   [U=900,P=1042]     PostStartProcess()            Post start process
Singularity lolcow_latest.sif:~> 
```

### Actual behavior

However, in `/vagrant` it fails to mount.

```
VERBOSE [U=0,P=1006]       print()                       Set messagelevel to: 5
DEBUG   [U=0,P=1006]       init()                        PIPE_EXEC_FD value: 7
VERBOSE [U=0,P=1006]       init()                        Container runtime
VERBOSE [U=0,P=1006]       is_suid()                     Check if we are running as setuid
DEBUG   [U=0,P=1006]       init()                        Overlay seems supported by kernel
DEBUG   [U=0,P=1006]       init()                        Drop privileges
DEBUG   [U=900,P=1006]     init()                        Read json configuration from pipe
DEBUG   [U=900,P=1006]     init()                        Set child signal mask
DEBUG   [U=900,P=1006]     init()                        Wait completion of stage1
VERBOSE [U=900,P=1011]     priv_escalate()               Get root privileges
DEBUG   [U=0,P=1011]       set_parent_death_signal()     Set parent death signal to 9
DEBUG   [U=0,P=1011]       prepare_stage()               Entering in stage 1
DEBUG   [U=900,P=1011]     set_parent_death_signal()     Set parent death signal to 9
VERBOSE [U=900,P=1011]     init()                        Spawn stage 1
VERBOSE [U=900,P=1011]     startup()                     Execute stage 1
DEBUG   [U=900,P=1011]     Stage()                       Entering stage 1
DEBUG   [U=900,P=1011]     Init()                        Entering image format intializer
DEBUG   [U=900,P=1011]     Init()                        Check for image format sif
DEBUG   [U=900,P=1011]     Init()                        sif format initializer returns: SIF magic not found
DEBUG   [U=900,P=1011]     Init()                        Check for image format sandbox
DEBUG   [U=900,P=1011]     Init()                        sandbox format initializer returns: not a directory image
DEBUG   [U=900,P=1011]     Init()                        Check for image format squashfs
DEBUG   [U=900,P=1006]     init()                        Create socketpair for master communication channel
DEBUG   [U=900,P=1006]     cleanup_fd()                  Check file descriptor /proc/self/fd/3 pointing to /vagrant/lolcow_latest.sif
DEBUG   [U=900,P=1006]     cleanup_fd()                  Check file descriptor /proc/self/fd/4 pointing to anon_inode:[eventpoll]
DEBUG   [U=900,P=1006]     cleanup_fd()                  Closing /proc/self/fd/4
VERBOSE [U=900,P=1006]     priv_escalate()               Get root privileges
VERBOSE [U=0,P=1006]       create_namespace()            Create mount namespace
DEBUG   [U=0,P=1006]       init()                        Create RPC socketpair for communication between stage 2 and RPC server
DEBUG   [U=0,P=1006]       set_terminal_control()        Pass terminal control to child
VERBOSE [U=0,P=1006]       init()                        Spawn master process
DEBUG   [U=0,P=1017]       set_parent_death_signal()     Set parent death signal to 9
VERBOSE [U=0,P=1017]       create_namespace()            Create mount namespace
DEBUG   [U=0,P=1017]       set_parent_death_signal()     Set parent death signal to 9
DEBUG   [U=0,P=1017]       prepare_stage()               Entering in stage 2
DEBUG   [U=900,P=1017]     set_parent_death_signal()     Set parent death signal to 9
VERBOSE [U=0,P=1020]       init()                        Spawn RPC server
VERBOSE [U=900,P=1006]     startup()                     Execute master process
DEBUG   [U=900,P=1006]     setupSessionLayout()          Attempting to use overlayfs (enable overlay = try)
DEBUG   [U=900,P=1006]     setupOverlayLayout()          Creating overlay SESSIONDIR layout
VERBOSE [U=0,P=1020]       startup()                     Serve RPC requests
DEBUG   [U=900,P=1006]     addRootfsMount()              Mount rootfs in read-only mode
DEBUG   [U=900,P=1006]     addRootfsMount()              Mounting block [squashfs] image: /vagrant/lolcow_latest.sif
DEBUG   [U=900,P=1006]     addKernelMount()              Checking configuration file for 'mount proc'
DEBUG   [U=900,P=1006]     addKernelMount()              Adding proc to mount list
VERBOSE [U=900,P=1006]     addKernelMount()              Default mount: /proc:/proc
DEBUG   [U=900,P=1006]     addKernelMount()              Checking configuration file for 'mount sys'
DEBUG   [U=900,P=1006]     addKernelMount()              Adding sysfs to mount list
VERBOSE [U=900,P=1006]     addKernelMount()              Default mount: /sys:/sys
DEBUG   [U=900,P=1006]     addDevMount()                 Checking configuration file for 'mount dev'
DEBUG   [U=900,P=1006]     addDevMount()                 Adding dev to mount list
VERBOSE [U=900,P=1006]     addDevMount()                 Default mount: /dev:/dev
DEBUG   [U=900,P=1006]     addHostMount()                Not mounting host file systems per configuration
VERBOSE [U=900,P=1006]     addBindsMount()               Found 'bind path' = /etc/localtime, /etc/localtime
VERBOSE [U=900,P=1006]     addBindsMount()               Found 'bind path' = /etc/hosts, /etc/hosts
DEBUG   [U=900,P=1006]     addHomeStagingDir()           Staging home directory (/home/vagrant) at /usr/local/var/singularity/mnt/session/home/vagrant
DEBUG   [U=900,P=1006]     addHomeMount()                Adding home directory mount [/usr/local/var/singularity/mnt/session/home/vagrant:/home/vagrant] to list using layer: overlay
DEBUG   [U=900,P=1006]     isLayerEnabled()              Using Layer system: overlay
DEBUG   [U=900,P=1006]     addTmpMount()                 Checking for 'mount tmp' in configuration file
VERBOSE [U=900,P=1006]     addTmpMount()                 Default mount: /tmp:/tmp
VERBOSE [U=900,P=1006]     addTmpMount()                 Default mount: /var/tmp:/var/tmp
DEBUG   [U=900,P=1006]     addScratchMount()             Not mounting scratch directory: Not requested
VERBOSE [U=900,P=1006]     addCwdMount()                 Default mount: /vagrant: to the container
DEBUG   [U=900,P=1006]     addLibsMount()                Checking for 'user bind control' in configuration file
DEBUG   [U=900,P=1006]     addResolvConfMount()          Adding /etc/resolv.conf to mount list
VERBOSE [U=900,P=1006]     addResolvConfMount()          Default mount: /etc/resolv.conf:/etc/resolv.conf
DEBUG   [U=900,P=1006]     addHostnameMount()            Skipping hostname mount, not virtualizing UTS namespace on user request
DEBUG   [U=900,P=1006]     create()                      Mount all
DEBUG   [U=900,P=1006]     mountGeneric()                Mounting tmpfs to /usr/local/var/singularity/mnt/session
DEBUG   [U=900,P=1006]     mountImage()                  Mounting loop device /dev/loop0 to /usr/local/var/singularity/mnt/session/rootfs
DEBUG   [U=900,P=1006]     CleanupContainer()            Cleanup container
FATAL   [U=900,P=1006]     Master()                      container creation failed: mount /proc/self/fd/3->/usr/local/var/singularity/mnt/session/rootfs error: can't mount image /proc/self/fd/3: failed to mount squashfs filesystem: input/output error
```

### Steps to reproduce behavior

Using the following Vagrant file:

```
Vagrant.configure("2") do |config|
  config.vm.box = "sylabs/singularity-3.1-ubuntu-bionic64"
  config.vm.box_version = "20190405.0.0"
end
```

```
vagrant up
vagrant ssh
cd /vagrant
singularity pull shub://GodloveD/lolcow
singularity --debut shell lolcow_latest.sif
```

Note that the same does not seem to be the case for the Centos7 box. 

I am using vagrant 2.2.4 and VirtualBox 6.0.6, on OSX Mojave 10.14.3.

Thank you.

**Comment by cclerget** [comment]
Hi @andersgs, this is a known issue with vagrant, as `/vagrant` is a virtual box filesystem, it doesn't work correctly with Singularity usage, you can't use singularity single image stored in this directory, you need to store them in another directory like vagrant home, /tmp or any other directory you have write permission.

**Comment by andersgs** [comment]
Thank you @cclerget. Weird that it works with the Centos7 box. But, because I need a larger VM size, the Ubuntu box is setup for easily extending the volume size. I wasn’t able to do the same with the Centos box. I’ll just have to copy the Singularity images in then.

**Comment by tashrifbillah** [comment]
@cclerget @andersgs 

This issue of `can't mount image /proc/self/fd/3: failed to mount squashfs filesystem: input/output error` is not specific to `/vagrant` folder only. I tried bind-mounting a different directory through `Vagrant` like below:

```
Vagrant.configure("2") do |config|
  config.vm.box = "sylabs/singularity-3.1-ubuntu-bionic64"
  config.vm.box_version = "20190405.0.0"

 config.vm.synced_folder "C:\\Users\tash\test_dir", "/home/vagrant"
end

```

Launch the box:

> vagrant up
> vagrant ssh

Then inside the `box`, I tried launching an image from the destination directory:

```
cd /home/vagrant/
singularity pull fsl_tbss.sif library://tbillah/collection/fsl_tbss:0.0.1
singularity shell --contain fsl_tbss.sif
```

And I ran into same error.

That means, we can't launch a singularity container from a bind-mounted volume through vagrant. We have to launch it from another directory. Then, whatever data the container processes, have to be copied back to the mounted directory so they persist on the host machine.
